File: Assets/Scripts/GrapplingHook.js
#pragma strict

import System.Collections.Generic;

var retractSpeed = 1.0;
var crosshairDistance = 2.5;
var offsetMagnitude : float = 0.1;
var minSeperation = 0.1;
var grappleable : LayerMask;
var solid : LayerMask;
var maxDistance : float = 20;

var crossTexture : Texture2D;
var hookNoise : AudioClip;
var hookNoise2 : AudioClip;

private var hook : GameObject;
private var rope : SpringJoint2D;
private var dragRope : DistanceJoint2D;
private var hookRenderer : Renderer;
private var line : LineRenderer;
private var controller : UnityStandardAssets._2D.PlatformerCharacter2D;
private var body : Rigidbody2D;
private var hookJoint : HingeJoint2D;
private var hookBody : Rigidbody2D;
private var crosshair : RectTransform;
private var crosshairRenderer : Canvas;
private var cameraScript : UnityStandardAssets._2D.Camera2DFollow;
private var hookAudio : AudioSource;
private var anim : Animator;
private var colliderBox : BoxCollider2D;

private var active = false;
private var desiredDistance : float;
private var controllerAim : float = -1.0;
private var controllerActive : boolean = false;
private var justBroken = false;
private var autoRegrapple = true;
private var lastGrappled : Transform;
private var cameraPlane : Plane[];

function Awake() {
	controller = gameObject.GetComponent(UnityStandardAssets._2D.PlatformerCharacter2D);
	cameraScript = Camera.main.GetComponent(UnityStandardAssets._2D.Camera2DFollow);
}

function Start() {
	hook = GameObject.Find("GrappleHook");
	rope = GetComponents.<SpringJoint2D>()[0];
	dragRope = GetComponents.<DistanceJoint2D>()[0];
	hookRenderer = hook.GetComponent(Renderer);
	line = GetComponent(LineRenderer);
	body = GetComponent(Rigidbody2D);
	hookJoint = hook.GetComponent(HingeJoint2D);
	hookBody = hook.GetComponent(Rigidbody2D);
	crosshair = GameObject.Find("Crosshair").GetComponent(RectTransform);
	crosshairRenderer = crosshair.gameObject.GetComponent(Canvas);
	hookAudio = hook.GetComponent(AudioSource);
	anim = GetComponent(Animator);
	colliderBox = transform.Find("Minimap").gameObject.GetComponent(BoxCollider2D);
	cameraPlane = GeometryUtility.CalculateFrustumPlanes(Camera.main);
	
	rope.connectedBody = hookBody;
	dragRope.connectedBody = hookBody;
}

function Vector3to2(v3 : Vector3) : Vector2 {return new Vector2(v3.x, v3.y);}

function GetMousePoint() : Vector2 {
	var v3 : Vector3 = Camera.main.ScreenToWorldPoint(Input.mousePosition);
	return new Vector2(v3.x, v3.y);
}

function GrappleRaycast() : RaycastHit2D {
	var angleToMouse : float;
	var mousePoint = GetMousePoint();
	if (mousePoint.y >= transform.position.y) {
		angleToMouse = Vector2.Angle(mousePoint - Vector3to2(transform.position), Vector2.right);
	} else {
		angleToMouse = 360 - Vector2.Angle(mousePoint - Vector3to2(transform.position), Vector2.right);
	}
	if (controllerActive) {angleToMouse = controllerAim;}
	var rotation = Quaternion.AngleAxis(angleToMouse, Vector3.forward) * Vector3.right;
	return Physics2D.Raycast(transform.position, rotation, maxDistance, solid);
}

function AngleToPlayer(point : Vector2, dir : Vector2) : float {
	var result = Vector2.Angle(point - transform.position, dir);
	if (point.x > transform.position.x) {result *= -1;}
	return result;
}

function BreakRope(noise : boolean) {
	active = false;
	rope.enabled = false;
	dragRope.enabled = false;
	hookJoint.enabled = false;
	hookBody.isKinematic = true;
	justBroken = true;
}

function GrappleInitial(toPoint : Vector2) {
	if (!justBroken || autoRegrapple) {
		var angleToPoint : float;
		//Get correct angle
		if (toPoint.y >= transform.position.y) {
			angleToPoint = Vector2.Angle(toPoint - Vector3to2(transform.position), Vector2.right);
		} else {
			angleToPoint = 360 - Vector2.Angle(toPoint - Vector3to2(transform.position), Vector2.right);
		}
		var rotation = Quaternion.AngleAxis(angleToPoint, Vector3.forward) * Vector3.right;
		var grapple = Physics2D.Raycast(transform.position, rotation, maxDistance, solid);
		
		//If I hit something not metal
		if (grapple.collider != null && (grappleable.value & (1 << grapple.transform.gameObject.layer)) > 0) { 
			//Move hook away from center a bit
			var variance = Vector2.zero;
			if (grapple.point.x > grapple.transform.position.x) {variance.x = offsetMagnitude;
			} else {variance.x = offsetMagnitude * -1;}
			if (grapple.point.y > grapple.transform.position.y) {variance.y = offsetMagnitude;
			} else {variance.y = offsetMagnitude * -1;}
			hook.transform.position = grapple.point + variance;
			lastGrappled = grapple.transform; //Save the grappled object
			
			if (grapple.rigidbody != null && !grapple.rigidbody.isKinematic) {
				hookJoint.connectedBody = grapple.rigidbody;
				hookJoint.connectedAnchor = grapple.transform.InverseTransformPoint(grapple.point);
				hookJoint.enabled = true;
				hookBody.isKinematic = false;
				
				dragRope.distance = Vector2.Distance(transform.position, hook.transform.position);
				dragRope.enabled = true;
				desiredDistance = dragRope.distance;
				hookAudio.PlayOneShot(hookNoise2, 0.5);
			} else {
				if (transform.position.y < hook.transform.position.y) {
					rope.enabled = true;
					rope.distance = Vector2.Distance(transform.position, hook.transform.position) - 1;
				} else if (body.velocity.x == 0) {
					rope.enabled = true;
					rope.distance = Vector2.Distance(transform.position, hook.transform.position);
				}
				desiredDistance = rope.distance;
				body.fixedAngle = false;
				hookAudio.PlayOneShot(hookNoise);
			}
			active = true;
		}
	}
}

function CheckRope() {
	var angle = AngleToPlayer(hook.transform.position, Vector2.up);
	var hit = Physics2D.Raycast(transform.position, 
		Quaternion.AngleAxis(angle, Vector3.forward) * Vector3.up, 
		Vector2.Distance(transform.position, hook.transform.position), solid);
	if (hit.collider != null && (hit.rigidbody == null || hit.rigidbody.isKinematic) && 
	Vector2.Distance(hit.point, hook.transform.position) > minSeperation) {
		if ((grappleable.value & (1 << hit.transform.gameObject.layer)) > 0) {
			if ((hit.transform.gameObject.layer == LayerMask.NameToLayer("GrappleLock")) && 
					(lastGrappled == hit.transform)) {

			} else {
				GrappleInitial(hit.point);
			}
		} else {
			BreakRope(true);
		}
	}
}

function GrappleUpdate() {
	desiredDistance -= retractSpeed;
	if (desiredDistance < 0) {desiredDistance = 0;}
	if (!hookJoint.enabled) {
		rope.distance = Mathf.Lerp(rope.distance, desiredDistance, Time.deltaTime);
		if (transform.position.y < hook.transform.position.y || 
			(Mathf.Abs(body.velocity.x) > (body.velocity.y * -1)) && (Mathf.Abs(body.velocity.y) > 2)) {
			rope.enabled = true;
		}
	}
	if (dragRope.enabled) {dragRope.distance = Mathf.Lerp(dragRope.distance, desiredDistance, Time.deltaTime);}
	CheckRope();
}

function FixedUpdate() {
	if (Input.GetButton("Grapple")) {
		if (active) {
			GrappleUpdate();
		} else {
			GrappleInitial(GetMousePoint());
		}
		if (Input.GetMouseButton(0) && controllerActive) {controllerActive = false;}
	} else {
		BreakRope(false);
		justBroken = false;
	}
	if ((controller.m_Grounded || controller.m_Tilted) && !body.fixedAngle && 
		(!active) && (body.velocity.magnitude < 1)) {
		if (controller.m_Tilted) {transform.position.y += 1;}
		transform.eulerAngles.z = 0;
		body.fixedAngle = true;
	}
	if (Input.GetKey(KeyCode.R) || Input.touchCount >= 3) {
		Respawn();
	}
	if (Input.GetKey(KeyCode.B)) {
		GameObject.Find("CratePink").transform.position = Vector3.left * 5;
	}
	
	/*if (!GeometryUtility.TestPlanesAABB(cameraPlane, colliderBox.bounds)) {
		Respawn();
	} */
}

function RenderHookLine() {
	if (active) {
		line.enabled = true;
		hookRenderer.enabled = true;
		line.SetPosition(0, transform.position);
		line.SetPosition(1, hook.transform.position);
	} else {
		line.enabled = false;
		hookRenderer.enabled = false;
	}
	anim.SetBool("Grappling", active);
}

function ControllerAimInput() {
	if ((Input.GetAxis("AimHorizontal") != 0) || (Input.GetAxis("AimVertical") != 0)) {
		if (Input.GetAxis("AimVertical") < 0) {
			controllerAim = Vector2.Angle(new Vector2(Input.GetAxis("AimHorizontal"), 
				Input.GetAxis("AimVertical")), Vector2.right);
		} else {
			controllerAim = 360 - Vector2.Angle(new Vector2(Input.GetAxis("AimHorizontal"), 
				Input.GetAxis("AimVertical")), Vector2.right);
		}
		controllerActive = true;
	}
	if (controllerActive) {
		crosshairRenderer.enabled = true;
		crosshair.position = transform.position + Vector3.back + (crosshairDistance * 
			(Quaternion.AngleAxis(controllerAim, Vector3.forward) * Vector3.right));
		Cursor.visible = false;
	} else {crosshairRenderer.enabled = false; Cursor.visible = true;}
}

function Update() {
	RenderHookLine();
	ControllerAimInput();
}

function Respawn() {
	transform.position = Vector3.zero;
	body.velocity = Vector3.zero;
	BreakRope(false);
}

File: Assets/Scripts/GrapplingHook.cs
using UnityEngine;
using System.Collections;

public class GrapplingHook : MonoBehaviour {

	public float retractSpeed = 2.0f;
	public float crosshairDistance = 2.5f;
	public float offsetMagnitude = 0.1f;
	public float minSeperation = 0.1f;
	public LayerMask grappleable;
	public LayerMask solid;
	public float maxDistance = 20f;
	public float armPosZ = 0.48f;
	
	public Texture2D crossTexture;
	public AudioClip hookNoise;
	public AudioClip hookNoise2;
	
	public Transform hook;
	private SpringJoint2D rope;
	private DistanceJoint2D dragRope;
	private Renderer hookRenderer;
	private UnityStandardAssets._2D.PlatformerCharacter2D controller;
	private Rigidbody2D body;
	private HingeJoint2D hookJoint;
	[HideInInspector]
	public Rigidbody2D hookBody;
	[HideInInspector]
	public RectTransform crosshair;
	private Canvas crosshairRenderer;
	[HideInInspector]
	public UnityStandardAssets._2D.Camera2DFollow cameraScript;
	private AudioSource hookAudio;
	private Animator anim;
	private Transform rightArm;
	private Transform rightHand;
	private Transform rightCoil;
	[HideInInspector]
	public Transform handUpper;
	[HideInInspector]
	public Transform handLower;

	[HideInInspector]
	public bool active = false;
	private float desiredDistance;
	private float controllerAim = -1.0f;
	private bool controllerActive = false;
	private bool justBroken = false;
	private bool autoRegrapple = true;
	[HideInInspector]
	public Transform lastGrappled;

	void Start() {
		controller = gameObject.GetComponent<UnityStandardAssets._2D.PlatformerCharacter2D>();
		rope = GetComponent<SpringJoint2D>();
		dragRope = GetComponent<DistanceJoint2D>();
		body = GetComponent<Rigidbody2D>();
		anim = GetComponent<Animator>();
		rightArm = transform.Find("robot/Arm_Right");
		rightHand = rightArm.Find("Arm_Front");
		rightCoil = GameObject.Find("ArmPart Seperate").transform;
	}

	public void StartTwo() {
		hookRenderer = hook.GetComponent<Renderer>();
		hookJoint = hook.GetComponent<HingeJoint2D>();
		hookAudio = hook.GetComponent<AudioSource>();
		crosshairRenderer = crosshair.gameObject.GetComponent<Canvas>();
		handUpper = hook.Find("Hand_Top");
		handLower = hook.Find("Hand_Bottom");
		GetComponent<SpringJoint2D>().connectedBody = hookBody;
		GetComponent<DistanceJoint2D>().connectedBody = hookBody;
	}

	Vector2 Vector3to2(Vector3 v3) {return new Vector2(v3.x, v3.y);}

	Vector2 GetMousePoint() {
		Vector3 mousePos = Input.mousePosition;
		mousePos.z = Mathf.Abs(Camera.main.transform.position.z);
		Vector3 v3 = Camera.main.ScreenToWorldPoint(mousePos);
		return new Vector2(v3.x, v3.y);
	}
	
	RaycastHit2D GrappleRaycast(Vector2 point) {
		float angle;
		if (point.y >= transform.position.y) {
			angle = Vector2.Angle(point - Vector3to2(transform.position), Vector2.right);
		} else {
			angle = 360f - Vector2.Angle(point - Vector3to2(transform.position), Vector2.right);
		}
		if (controllerActive) {angle = controllerAim;}
		Vector3 rotation = Quaternion.AngleAxis(angle, Vector3.forward) * Vector3.right;
		return Physics2D.Raycast(transform.position, rotation, maxDistance, solid);
	}
	
	float AngleToPoint(Vector2 point, bool flip) {
		Vector2 side = Vector2.right; if (!flip) {side *= -1f;}
		float result = Vector2.Angle(point - Vector3to2(transform.position), side);
		if (point.y < transform.position.y) {result = 360f - result;}
		return result; 
	}
	
	float AngleToPlayer(Vector2 point, Vector2 dir) {
		float result = Vector2.Angle(point - Vector3to2(transform.position), dir);
		if (point.x > transform.position.x) {result *= -1f;}
		return result;
	}

	void BreakRope(bool noise) {
		active = false;
		rope.enabled = false;
		dragRope.enabled = false;
		hookJoint.enabled = false;
		hookBody.isKinematic = true;
		justBroken = true;
	}

	void GrappleInitial(Vector2 toPoint) {
		if (!justBroken || autoRegrapple) {
			RaycastHit2D grapple = GrappleRaycast(toPoint);

			if (grapple.collider != null && (grappleable.value & (1 << grapple.transform.gameObject.layer)) > 0) { 
				Vector2 variance = Vector2.zero;
				if (grapple.point.x > grapple.transform.position.x) {variance.x = offsetMagnitude;
				} else {variance.x = offsetMagnitude * -1f;}
				if (grapple.point.y > grapple.transform.position.y) {variance.y = offsetMagnitude;
				} else {variance.y = offsetMagnitude * -1f;}
				hook.position = grapple.point + variance;
				hook.position += Vector3.back*armPosZ;
				lastGrappled = grapple.transform; 
				
				if (grapple.rigidbody != null && !grapple.rigidbody.isKinematic) {
					hookJoint.connectedBody = grapple.rigidbody;
					hookJoint.connectedAnchor = grapple.transform.InverseTransformPoint(grapple.point);
					hookJoint.enabled = true;
					hookBody.isKinematic = false;
					
					dragRope.distance = Vector2.Distance(transform.position, hook.position);
					dragRope.enabled = true;
					desiredDistance = dragRope.distance;
					hookAudio.PlayOneShot(hookNoise2, 0.5f);
				} else {
					rope.enabled = true;
					rope.distance = Vector2.Distance(transform.position, hook.position) - 1f;
					
					desiredDistance = rope.distance;
					body.fixedAngle = false;
					hookAudio.PlayOneShot(hookNoise);
				}
				active = true;
			}
		}
	}

	public void GrappleInitialSync() {
		Transform blocks = GameObject.Find("World").transform;
		float closestDist = Mathf.Infinity;
		foreach (Transform block in blocks) {
			if ((grappleable.value & (1 << block.gameObject.layer)) > 0) {
				float dist = Vector2.Distance(hook.position, block.position);
				if (dist < closestDist) {
					closestDist = dist;
					lastGrappled = block;
				}
			}
		}
		Rigidbody2D lastRigidbody = lastGrappled.GetComponent<Rigidbody2D>();
		if (lastRigidbody != null && !lastRigidbody.isKinematic) {
			hookJoint.connectedBody = lastRigidbody;
			hookJoint.connectedAnchor = lastGrappled.InverseTransformPoint(hook.position);
			hookJoint.enabled = true;
			hookBody.isKinematic = false;
			
			dragRope.distance = Vector2.Distance(transform.position, hook.position);
			dragRope.enabled = true;
			desiredDistance = dragRope.distance;
			hookAudio.PlayOneShot(hookNoise2, 0.5f);
		} else {
			rope.enabled = true;
			rope.distance = Vector2.Distance(transform.position, hook.position) - 1f;
			
			desiredDistance = rope.distance;
			body.fixedAngle = false;
			hookAudio.PlayOneShot(hookNoise);
		}
		active = true;
	}

	void CheckRope() {
		float angle = AngleToPlayer(hook.position, Vector2.up);
		RaycastHit2D hit = Physics2D.Raycast(transform.position, 
		                            Quaternion.AngleAxis(angle, Vector3.forward) * Vector3.up, 
		                            Vector2.Distance(transform.position, hook.position), solid);
		if (hit.collider != null && (hit.rigidbody == null || hit.rigidbody.isKinematic) && 
		    Vector2.Distance(hit.point, hook.position) > minSeperation) {
			if ((grappleable.value & (1 << hit.transform.gameObject.layer)) > 0) {
				if (!((hit.transform.gameObject.layer == LayerMask.NameToLayer("GrappleLock")) && (lastGrappled == hit.transform))) {
					GrappleInitial(hit.point);}
			} else {
				BreakRope(true);
			}
		}
	}

	void GrappleUpdate() {
		desiredDistance -= retractSpeed;
		if (desiredDistance < 0) {desiredDistance = 0;}
		if (!hookJoint.enabled) {rope.distance = Mathf.Lerp(rope.distance, desiredDistance, Time.deltaTime);}
		if (dragRope.enabled) {dragRope.distance = Mathf.Lerp(dragRope.distance, desiredDistance, Time.deltaTime);}
		CheckRope();
	}

	void FixedUpdate() {
		if (Input.GetButton("Grapple") && Input.touchCount < 2) {
			if (active) {
				GrappleUpdate();
			} else {
				GrappleInitial(GetMousePoint());
			}
			if (Input.GetMouseButton(0) && controllerActive) {controllerActive = false;}
		} else {BreakRope(false); justBroken = false;}
		if ((controller.m_Grounded || controller.m_Tilted) && !body.fixedAngle && (!active) && (body.velocity.magnitude < 1f)) {
			if (controller.m_Tilted) {transform.position += Vector3.up;}
			//transform.eulerAngles.z = 0f;
			transform.rotation = Quaternion.identity;
			body.fixedAngle = true;
		}
		if (Input.GetKey(KeyCode.R) || Input.touchCount >= 3) {Respawn();}
	}

	public void Respawn() {
		transform.position = Vector3.zero;
		body.velocity = Vector3.zero;
		BreakRope(false);
	}

	void RenderHookLine() {
		if (active) {
			hookRenderer.enabled = true;
			handUpper.gameObject.SetActive(true);
			handLower.gameObject.SetActive(true);
			
			Vector3 between = hook.position - rightArm.position;
			float distance = between.magnitude;
			rightCoil.localScale = new Vector3(rightCoil.localScale.x, rightCoil.localScale.y, distance);// * 0.56448285185f);
			rightCoil.position = rightArm.position + (between / 2f);
			rightCoil.LookAt(hook.position);
			rightCoil.GetComponent<MeshRenderer>().enabled = true;
		} else {
			hookRenderer.enabled = false;
			handUpper.gameObject.SetActive(false);
			handLower.gameObject.SetActive(false);
			
			rightCoil.localPosition = Vector3.right * 0.151f;
			rightCoil.rotation = Quaternion.identity;
			rightCoil.localScale = new Vector3(0.15f, 0.15f, 0.15f);
			rightCoil.GetComponent<MeshRenderer>().enabled = false;
		}
		anim.SetBool("Grappling", active);
	}

	void ControllerAimInput() {
		if ((Input.GetAxis("AimHorizontal") != 0f) || (Input.GetAxis("AimVertical") != 0f)) {
			if (Input.GetAxis("AimVertical") < 0f) {
				controllerAim = Vector2.Angle(new Vector2(Input.GetAxis("AimHorizontal"), Input.GetAxis("AimVertical")), Vector2.right);
			} else {
				controllerAim = 360f - Vector2.Angle(new Vector2(Input.GetAxis("AimHorizontal"), Input.GetAxis("AimVertical")), Vector2.right);
			}
			controllerActive = true;
		}
		if (controllerActive) {
			crosshairRenderer.enabled = true;
			crosshair.position = transform.position + Vector3.back + (crosshairDistance * (Quaternion.AngleAxis(controllerAim, Vector3.forward) * Vector3.right));
			Cursor.visible = false;
		} else {crosshairRenderer.enabled = false; Cursor.visible = true;}
	}

	void Update() {
		RenderHookLine();
		ControllerAimInput();
		
		float armAngle;
		if (active) {
			armAngle = AngleToPoint(Vector3to2(hook.position), true);
			rightHand.gameObject.SetActive(false);
		} else {
			armAngle = AngleToPoint(GetMousePoint(), controller.m_FacingRight);
			rightHand.gameObject.SetActive(true);
		}
		Vector3 initRotation = rightArm.rotation.eulerAngles;
		initRotation.z = armAngle;
		rightArm.rotation = Quaternion.Euler(initRotation);
		hook.rotation = Quaternion.Euler(initRotation);
	}
}